File: python/src/thunder_book/ch05/iterative_deepening.py
import fire

from thunder_book.ch05.alpha_beta import alpha_beta_action
from thunder_book.ch05.average_score import average_score
from thunder_book.ch05.maze_state import AlternateMazeState, MazeParams
from thunder_book.ch05.time_keeper import TimeKeeper


def iterative_deepening_action(state: AlternateMazeState, time_threshold: int):
    time_keeper = TimeKeeper(time_threshold)
    best_action = 0
    # depth を大きくすると (e.g. 1000) なぜかバグる, 本来 END_TURN までやれば良いよね？
    for depth in range(1, state.params.end_turn * 2):
        best_action = alpha_beta_action(state, depth=depth, time_keeper=time_keeper)
        if time_keeper.is_time_over():
            # print(f"depth: {depth} with time {time_keeper.time_threshold}")
            break
    return best_action


def compare_iterative_deepening(num_games: int = 100, a: int = 10, b: int = 1):
    print(f"compare iterative deepening {a}ms vs. {b}ms")
    params = MazeParams(width=5, height=5, end_turn=10)
    iterative_deepening_a = lambda state: iterative_deepening_action(state, a)
    iterative_deepening_b = lambda state: iterative_deepening_action(state, b)
    win_rate = average_score(
        num_games, (iterative_deepening_a, iterative_deepening_b), params=params
    )
    print(f"win rate of iterative deepening {a}ms vs. {b}ms: {win_rate:.2f}")


if __name__ == "__main__":
    fire.Fire(compare_iterative_deepening)

File: rust/src/ch05/iterative_deepening.rs
use std::sync::Arc;

use crate::base::alternate::{AlternateState, Evaluatable};
use crate::base::is_done;
use crate::ch05::alpha_beta;
use crate::ch05::maze_state;

// tracks the final result of the deepening
#[derive(Debug)]
struct DeepeningResult {
    action: usize,
    _depth: usize,
}

// (state, depth) -> action
// type Deepenable = dyn Fn(&maze_state::AlternateMazeState, usize) -> usize;

// do alpha-beta, keep deeper until time is up
// it doesn't early stop when time is up during iteration
// but won't use the result so isn't cheating.
// 早めに終わる実装をやってもいいけど alpha-beta には手を入れたくないのでこのままで
// is_time_up decorator を作って各loopも関数にして、見たいのでやればできる気はする
fn iterative_deepening_action<T: AlternateState + Evaluatable>(
    state: &T,
    time_threshold_ms: u64,
    // todo: action: Deepenable にして alpha beta 側に実装する
) -> DeepeningResult {
    let mut best: Option<usize> = None;
    let mut depth = 1;

    let mut is_time_up = is_done::time_stopper(time_threshold_ms);
    loop {
        // call first because timer starts from first call
        if is_time_up() {
            break;
        }
        // todo: action_func(&state) にする
        let action = alpha_beta::alpha_beta_arc(depth)(state);
        best = Some(action);
        depth += 1;
    }

    DeepeningResult {
        action: best.unwrap(),
        _depth: depth,
    }
}

pub fn iterative_deepening_action_arc<T: AlternateState + Evaluatable>(
    time_threshold_ms: u64,
) -> maze_state::ActionFunc<T> {
    Arc::new(move |state| -> usize {
        iterative_deepening_action(state, time_threshold_ms).action
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::base::alternate::MazeParams;

    fn setup() -> maze_state::AlternateMazeState {
        let params = MazeParams {
            height: 3,
            width: 3,
            end_turn: 4,
        };
        maze_state::AlternateMazeState::new(0, params)
    }

    #[test]
    fn test_iterative_deepening_arc() {
        let state = setup();
        let action = iterative_deepening_action_arc(1)(&state);
        assert_eq!(action, 0);
    }

    #[test]
    fn test_iterative_deepening() {
        let state = setup();
        let short = iterative_deepening_action(&state, 1);
        let long = iterative_deepening_action(&state, 2);

        assert!(short._depth < long._depth);
        // assert_eq!(short._depth, long._depth);
        assert_eq!(short.action, 0);
        assert_eq!(long.action, 0);
    }
}
